homeassistant/components/totalconnect/__init__.py

```diff
@@ -3,42 +3,36 @@
 import logging
 
 from total_connect_client import TotalConnectClient
-import voluptuous as vol
 
 from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
 from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
 from homeassistant.core import HomeAssistant
-import homeassistant.helpers.config_validation as cv
 
-from .const import DOMAIN
+from .const import CONF_USERCODES, DOMAIN
 
 _LOGGER = logging.getLogger(__name__)
 
 PLATFORMS = ["alarm_control_panel", "binary_sensor"]
 
-CONFIG_SCHEMA = vol.Schema(
-    {
-        DOMAIN: vol.Schema(
-            {
-                vol.Required(CONF_USERNAME): cv.string,
-                vol.Required(CONF_PASSWORD): cv.string,
-            }
-        )
-    },
-    extra=vol.ALLOW_EXTRA,
-)
-
 
 async def async_setup(hass: HomeAssistant, config: dict):
     """Set up by configuration file."""
     if DOMAIN not in config:
         return True
 
+    config_data = {}
+    config_data[CONF_USERNAME] = config[DOMAIN].get(CONF_USERNAME)
+    config_data[CONF_PASSWORD] = config[DOMAIN].get(CONF_PASSWORD)
+
+    if CONF_USERCODES in config[DOMAIN]:
+        config_data[CONF_USERCODES] = dict(config[DOMAIN][CONF_USERCODES])
+    else:
+        config_data[CONF_USERCODES] = {}
+        _LOGGER.error("TotalConnect configuration is missing usercodes")
```

Comment:
User codes is a new feature and wasn't part of the config schema. Why should we log an error? We know that this will always be the case for the config yaml.

We should just have a check in `async_setup_entry` that checks for missing user codes. If the user codes are missing we should fail the entry set up and start a new config flow and let the user authenticate again. Add a reauth step.

Reply:
Sorry the log entry was to help me debug and I forgot to remove it.  Given we are now removing YAML as an option, I simplified async_setup.